File: code/python/ch17/practice/practice_02_merge_k_sorted.py
"""
Practice 2: Merge K Sorted Arrays
=====================================
Chapter 17: Heaps & Priority Queues — The VIP Line

PROBLEM
-------
Given K sorted arrays, merge them into one sorted array.

INPUT FORMAT
------------
The function receives a list of sorted lists.

OUTPUT FORMAT
-------------
A single sorted list containing all elements.

CONSTRAINTS
-----------
- 0 <= K <= 100
- 0 <= length of each array <= 1000
- -10^9 <= elements <= 10^9

EXAMPLES
--------
Input: [[1, 4, 7], [2, 5, 8], [3, 6, 9]]
Output: [1, 2, 3, 4, 5, 6, 7, 8, 9]

Input: [[1, 3, 5], [2, 4, 6]]
Output: [1, 2, 3, 4, 5, 6]

Input: [[], [1]]
Output: [1]

HINT
----
Use a min-heap of size K. Push (value, array_index, element_index)
tuples. Pop the minimum, then push the next element from the same array.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(arrays: list[list[int]]) -> list[int]:
    """Merge K sorted arrays into one sorted array."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    import json
    arrays = json.loads(input().strip())
    print(solve(arrays))

File: code/cpp/ch17/practice/practice_02_merge_k_sorted.cpp
/*
 * Practice 2: Merge K Sorted Arrays
 * ====================================
 * Chapter 17: Heaps & Priority Queues — The VIP Line
 *
 * PROBLEM:
 *   Given K sorted arrays, merge them into one sorted array.
 *
 * EXAMPLES:
 *   solve({{1,4,7},{2,5,8},{3,6,9}})  -> {1,2,3,4,5,6,7,8,9}
 *   solve({{1,3,5},{2,4,6}})          -> {1,2,3,4,5,6}
 *
 * CONSTRAINTS:
 *   - 0 <= K <= 100
 *   - 0 <= arrays[i].size() <= 10^4
 *   - Each array is sorted in ascending order
 *
 * INSTRUCTIONS:
 *   Replace the body of solve() with your solution.
 */

#include <algorithm>
#include <iostream>
#include <queue>
#include <tuple>
#include <vector>
using namespace std;

vector<int> solve(vector<vector<int>> arrays) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int k;
    cin >> k;
    vector<vector<int>> arrays(k);
    for (int i = 0; i < k; i++) {
        int n;
        cin >> n;
        arrays[i].resize(n);
        for (int j = 0; j < n; j++) cin >> arrays[i][j];
    }
    vector<int> result = solve(arrays);
    for (int i = 0; i < (int)result.size(); i++) {
        if (i) cout << " ";
        cout << result[i];
    }
    cout << endl;
    return 0;
}
